include/nuttx/efuse/efuse.h

```diff
@@ -0,0 +1,294 @@
+ * Return:      Zero (OK) on success.  Minus one will be returned on failure
+ *              with the errno value set appropriately.
+ */
+
+#define EFUSEIOC_READ_BLOCK           _EFUSEIOC(0x0008)
+
+/* Command:     EFUSEIOC_WRITE_BLOCK
+ * Description: Write a key to efuse's block.
+ * Arguments:   A structure containing the block number, the src_key pointer,
+ *              the offset in bits and the amount of bits to write.
+ * Return:      Zero (OK) on success.  Minus one will be returned on failure
+ *              with the errno value set appropriately.
+ */
+
+#define EFUSEIOC_WRITE_BLOCK          _EFUSEIOC(0x0009)
+
+/* Command:     EFUSEIOC_BATCH_WRITE_BEGIN
+ * Description: Start a batch operation, the efuse bits will be burned after
+ *              a batch write commit operation.
+ * Arguments:   None
+ * Return:      Zero (OK) on success.  Minus one will be returned on failure
+ *              with the errno value set appropriately.
+ */
+
+#define EFUSEIOC_BATCH_WRITE_BEGIN    _EFUSEIOC(0x000a)
+
+/* Command:     EFUSEIOC_BATCH_WRITE_CANCEL
+ * Description: Cancel a started batch write operation in progress.
+ * Arguments:   None
+ * Return:      Zero (OK) on success.  Minus one will be returned on failure
+ *              with the errno value set appropriately.
+ */
+
+#define EFUSEIOC_BATCH_WRITE_CANCEL   _EFUSEIOC(0x000b)
+
+/* Command:     EFUSEIOC_BATCH_WRITE_COMMIT
+ * Description: Commit a batch operation that was in progress.
+ * Arguments:   None
+ * Return:      Zero (OK) on success.  Minus one will be returned on failure
+ *              with the errno value set appropriately.
+ */
+
+#define EFUSEIOC_BATCH_WRITE_COMMIT   _EFUSEIOC(0x000c)
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+/* Structure eFuse field */
+
+struct efuse_desc_s
+{
+  uint16_t   efuse_block;    /* Block of eFuse */
+  uint16_t   bit_start;      /* Start bit [0..65535] */
+  uint16_t   bit_count;      /* Length of bit field [1..-] */
+};
+
+/* Type definition for an eFuse field */
+
+typedef struct efuse_desc_s efuse_desc_t;
+
+/* Structure range address by blocks */
+
+typedef struct
+{
+  uint32_t start;
+  uint32_t end;
+} efuse_range_addr_t;
+
+/* Structs with the parameters passed to the IOCTLs */
+
+struct efuse_par
+{
+  uint16_t block;
+  uint16_t reg;
+  size_t   bit_offset;
+  size_t   bit_size;
+  uint8_t *data;
+};
+
+struct efuse_par_id
+{
+  efuse_desc_t **field;
+  size_t  size;
+  uint8_t *data;
+};
+
+/* This structure provides the "lower-half" driver operations available to
+ * the "upper-half" driver.
+ */
+
+struct efuse_lowerhalf_s;
+struct efuse_ops_s
+{
+  /* Required methods *******************************************************/
+
+  /* Read an EFUSE bit */
+
+  CODE int (*read_bit)(FAR struct efuse_lowerhalf_s *lower,
+                       const efuse_desc_t *field[],
```

Comment:
why need an array of efuse_desc_t

Reply:
I think that the original idea of passing an array of efuse_desc_t was to be able to read/write multiples fields with a single call, but it is not used in my driver neither in the IDF driver. I will modify it to pass a single structure.

Reply:
@xiaoxiang781216 I decided to keep the original array of efuse_desct_t, it brings more flexibility for fields definition. For example, we can define the MAC Addresses separating for bytes to keep a single stream:

```
static const efuse_desc_t MAC_FACTORY[] =
{
  {
    72, 8    /* Factory MAC addr [0], */
  },
  {
    64, 8    /* Factory MAC addr [1], */
  },
  {
    56, 8    /* Factory MAC addr [2], */
  },
  {
    48, 8    /* Factory MAC addr [3], */
  },
  {
    40, 8    /* Factory MAC addr [4], */
  },
  {
    32, 8    /* Factory MAC addr [5], */
  },
};


const efuse_desc_t *ESP_EFUSE_MAC_FACTORY[] =
{
  &MAC_FACTORY[0],      /* Factory MAC addr [0] */
  &MAC_FACTORY[1],      /* Factory MAC addr [1] */
  &MAC_FACTORY[2],      /* Factory MAC addr [2] */
  &MAC_FACTORY[3],      /* Factory MAC addr [3] */
  &MAC_FACTORY[4],      /* Factory MAC addr [4] */
  &MAC_FACTORY[5],      /* Factory MAC addr [5] */
  NULL
};


static const efuse_desc_t MAC_CUSTOM[] =
{
  {
    776, 48    /* Custom MAC */
  },
};

const efuse_desc_t *ESP_EFUSE_MAC_CUSTOM[] =
{
  &MAC_CUSTOM[0],       /* Custom MAC */
  NULL
};
```

Reply:
But I removed things that were ESP32 specific, like the idea of separating efuses by blocks, now efuse is just a linear stream of bits. All we need to define are the fields with each efuse offset and its bit size.

Reply:
Since efuse_desc_t is very flexible, doesn't this mean we just need two callback(read/write with efuse_desc_t array) now?

Reply:
Yes, exactly it what I did now! I removed the read/write block and read/write reg as you suggested! ;-)